File: WoofAndroid/design-system/src/main/kotlin/com/perrystreet/woof/designsystem/atomic/atoms/textfield/AtomTextField.kt
package com.perrystreet.woof.designsystem.atomic.atoms.textfield

import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.text.BasicTextField
import androidx.compose.foundation.text.KeyboardActions
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.SolidColor
import androidx.compose.ui.text.input.ImeAction
import com.perrystreet.woof.designsystem.atomic.atoms.text.roles.TextColorRole
import com.perrystreet.woof.designsystem.atomic.atoms.text.roles.TextFontRole
import com.perrystreet.woof.designsystem.theme.Theme

@Composable
fun AtomTextField(
    text: String,
    placeholder: String,
    onTextChange: (String) -> Unit,
    onSubmit: () -> Unit,
    textFontRole: TextFontRole,
    colorRole: TextColorRole,
    placeholderColorRole: TextColorRole,
    modifier: Modifier = Modifier,
) {
    val textStyle = textFontRole.textStyle()
    BasicTextField(
        value = text,
        onValueChange = onTextChange,
        modifier = modifier,
        textStyle = textStyle.copy(color = colorRole.color()),
        singleLine = true,
        keyboardOptions = KeyboardOptions(imeAction = ImeAction.Send),
        keyboardActions = KeyboardActions(onSend = { onSubmit() }),
        cursorBrush = SolidColor(Theme.colors.primary),
        decorationBox = { innerTextField ->
            Box {
                when (text.isEmpty()) {
                    true -> Text(
                        text = placeholder,
                        style = textStyle,
                        color = placeholderColorRole.color(),
                        maxLines = 1,
                    )
                    false -> Unit
                }
                innerTextField()
            }
        },
    )
}

File: WoofIOS/SwiftPackages/DesignSystem/Sources/DesignSystem/Atomic/Atoms/TextField/AtomTextField.swift
import SwiftUI

public struct AtomTextField: View {
    @Environment(\.theme) private var theme

    private let text: String
    private let placeholder: String
    private let onTextChange: (String) -> Void
    private let onSubmit: () -> Void
    private let textFontRole: TextFontRole
    private let colorRole: TextColorRole
    private let placeholderColorRole: TextColorRole

    public init(
        text: String,
        placeholder: String,
        onTextChange: @escaping (String) -> Void,
        onSubmit: @escaping () -> Void,
        textFontRole: TextFontRole,
        colorRole: TextColorRole,
        placeholderColorRole: TextColorRole
    ) {
        self.text = text
        self.placeholder = placeholder
        self.onTextChange = onTextChange
        self.onSubmit = onSubmit
        self.textFontRole = textFontRole
        self.colorRole = colorRole
        self.placeholderColorRole = placeholderColorRole
    }

    public var body: some View {
        TextField(
            "",
            text: Binding(get: { text }, set: onTextChange),
            prompt: Text(placeholder).foregroundStyle(placeholderColorRole.color(from: theme))
        )
        .font(textFontRole.font(from: theme))
        .foregroundStyle(colorRole.color(from: theme))
        .tint(theme.colors.primary)
        .submitLabel(.send)
        .onSubmit(onSubmit)
    }
}
